Propagate NaNs in the CPU min and max operators

adamreeve opened the pull request:
### Description

Propagates NaN values in the min and max operators so that min or max with a NaN in either input always produces NaN.

### Motivation and Context

Fixes #21455

skottmckay commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

skottmckay commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

skottmckay commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

skottmckay commented:
Should there be a test where the NaN is the scalar to check it is propagated throughout the broadcast?

e.g. input {2,2} with no NaN and {1} with a NaN should result in all NaN in the output IIUC

adamreeve commented:
@microsoft-github-policy-service agree company="G-Research"

adamreeve commented:
> Should there be a test where the NaN is the scalar to check it is propagated throughout the broadcast?

Good idea thanks, I've added those tests now and fixed the formatting errors.

skottmckay commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

skottmckay commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

skottmckay commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

adamreeve commented:
It looks like there are some issues with the asan runs and the MLFloat16 tests: 
* https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1447218&view=logs&j=39f143cf-c993-54bc-2606-bcdaa7decea3&t=9e0c99a9-dfb0-59b2-bc39-311770508d51&l=6413
* https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1447214&view=logs&j=2f248fe9-fc97-5fc1-8ff4-e568b41d705c&t=4ca5b659-f5e6-52ee-9d8f-90ebda7fd6f8&l=9780

And the web builds also seem to have a problem with these new tests: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1447211&view=logs&j=990616d7-1f75-5fe5-5c67-c84a39482fba&t=67a495a8-8444-507d-6780-4eb1ddef3707&l=14255

The same problem was run into in https://github.com/microsoft/onnxruntime/pull/19984#issuecomment-2010471172, where there weren't any changes to `element_wise_ops.cc` but there were tests added for handling NaNs with MLFloat16.

I'm not sure what could be causing this, it seems like a bug in Eigen rather than onnxruntime at first glance. The best way forward is probably to do the same as before and revert the MLFloat16 related changes for now, and I can make a new issue to follow up on this.

skottmckay commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

skottmckay commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

skottmckay commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline
